src/wallet/wallet.cpp

```diff
@@ -1147,6 +1153,50 @@ void CWallet::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) {
 
 
 
+void CWallet::BlockUntilSyncedToCurrentChain() {
+    const CBlockIndex* initialChainTip;
+    {
+        LOCK(cs_main);
+        initialChainTip = chainActive.Tip();
+    }
+    AssertLockNotHeld(cs_main);
+    AssertLockNotHeld(cs_wallet);
+    std::unique_lock<std::mutex> lock(lastBlockProcessedMutex);
+
+    assert(lastBlockProcessed);
+
+    auto pred = [this, initialChainTip] {
+            if (this->lastBlockProcessed == initialChainTip) {
+                return true;
+            }
+            if (this->lastBlockProcessed->GetAncestor(initialChainTip->nHeight) == initialChainTip) {
+                return true;
+            }
+            // Catch the race condition where the wallet may have caught up and
+            // moved past initialChainTip through a reorg before we could get
+            // lastBlockProcessedMutex.
+            // This should be exceedingly rare in regular usage, so potentially
+            // eating 100ms to retry this lock should be fine (not TRY_LOCKing
+            // here would be a lock inversion against lastBlockProcessedMutex)
+            TRY_LOCK(cs_main, mainLocked);
+            if (mainLocked) {
+                if (this->lastBlockProcessed == chainActive.Tip()) {
+                    return true;
+                }
+                // If the user called invalidateblock our wait here might block
+                // forever, so we check if we're ahead of the tip (a state
+                // which should otherwise never be exposed outside of validation)
+                return this->lastBlockProcessed->nChainWork > chainActive.Tip()->nChainWork;
```

Comment:
In commit "Add CWallet::BlockUntilSyncedToCurrentChain()"

I don't understand why we would wait forever without this check. Does InvalidateBlock not trigger notifications that would lead to lastBlockProcessed being updated? And if it doesn't, shouldn't this just be fixed so the right notifications are sent?